### PR TITLE
Support long conversion to float for 0 and 1

### PR DESCRIPTION
**Pull request by serena-ruan**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/serena-ruan/mlflow/pull/13036?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/13036/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 13036
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?
0 should be able to convert from long type to float/double type

### How is this PR tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/models/utils.py [hunk 1]
```diff
@@ -821,6 +821,17 @@ def _enforce_mlflow_datatype(name, values: pd.Series, t: DataType):
     if is_upcast:
         return values.astype(numpy_type, errors="raise")
     else:
+        # support converting long -> float/double for 0 values
+        def all_zeros(xs):
+            return all(pd.isnull(x) or x == 0 for x in xs)
```
Comment (mingyu89) [inline mlflow/models/utils.py:826]: What will be the value if the parameter is not provided?

**Review by mingyu89 — APPROVED** [review]
Thank you for the quick fix! Can we do a quick test to make sure it's working as expected?(It could be post merging)

File: mlflow/models/utils.py (hunk 1 above)
Comment (serena-ruan) [reply mlflow/models/utils.py:826]: If the field is optional, then this field just becomes null when casting to pandas Series, and it should pass the check. Updated the test

File: mlflow/models/utils.py [hunk 2]
```diff
@@ -821,6 +821,17 @@ def _enforce_mlflow_datatype(name, values: pd.Series, t: DataType):
     if is_upcast:
         return values.astype(numpy_type, errors="raise")
     else:
+        # support converting long -> float/double for 0 values
+        def all_zero_or_ones(xs):
+            return all(pd.isnull(x) or x in [0, 1] for x in xs)
+
+        if (
+            values.dtype == np.int64
```
Comment (WeichenXu123) [inline mlflow/models/utils.py:829]: int16 / int32 ?

File: mlflow/models/utils.py [hunk 3]
```diff
@@ -821,6 +821,17 @@ def _enforce_mlflow_datatype(name, values: pd.Series, t: DataType):
...
+            and numpy_type in (np.float32, np.float64)
```
Comment (WeichenXu123) [inline mlflow/models/utils.py:830]: np.float16  ?

File: mlflow/models/utils.py [hunk 4]
```diff
@@ -821,6 +821,17 @@ def _enforce_mlflow_datatype(name, values: pd.Series, t: DataType):
     if is_upcast:
         return values.astype(numpy_type, errors="raise")
     else:
+        # support converting long -> float/double for 0 values
+        def all_zero_or_ones(xs):
+            return all(pd.isnull(x) or x in [0, 1] for x in xs)
```
Comment (daniellok-db) [inline mlflow/models/utils.py:826]: if we want to do something like all int32s
```suggestion
        # support converting int32 -> float/double for 0 values
        def all_zero_or_ones(xs):
            return all(pd.isnull(x) or x < 2 ** 31 for x in xs)
```

Comment (daniellok-db) [reply mlflow/models/utils.py:826]: but i guess it's probably safer to just keep 0 or 1

Comment (WeichenXu123) [reply mlflow/models/utils.py:826]: shouldn't be `abs(x) < 2 ** 31` ?

File: mlflow/models/utils.py (hunk 2 above)
Comment (WeichenXu123) [reply mlflow/models/utils.py:829]: is it possible to be uint64 ?

File: mlflow/models/utils.py (hunk 4 above)
Comment (WeichenXu123) [reply mlflow/models/utils.py:826]: another issue is it uses for loop which is very slow,( it causes big overhead in enforcing data type), and it has many other similar for loop in  enforcing dat code, can we improve it ? (can be in follow-up pr)

like using numpy batch operation:
`(np.abs(xs) < 2**31).all()`

Comment (WeichenXu123) [reply mlflow/models/utils.py:826]: with vectorization:
```
(np.abs(xs) < 2**31).all()
```
without vectorization:
```
all(x < 2 ** 31 for x in xs)  # it iterates elements one by one
```

See https://www.programiz.com/python-programming/numpy/vectorization for more details of why vectorization making it faster.

Comment (WeichenXu123) [reply mlflow/models/utils.py:826]: can be a follow-up work though, we have many places to do the similar change

File: mlflow/models/utils.py (hunk 3 above)
Comment (serena-ruan) [reply mlflow/models/utils.py:830]: We don't have such type in Datatype, it's just float or double.

**Review by daniellok-db — APPROVED** [review]

**Pull request merged**